File: plugins/cartim/models/option.py
from database.db import db
from flask_mongoengine import BaseQuerySet
import datetime
from slugify import slugify, Slugify, UniqueSlugify
from flask_mongoengine.wtf import model_form


class Option(db.Document):
    sort = db.IntField(required=False, default=0)
    name = db.StringField(required=True, unique=True)
    display_name = db.ListField(required=False, default=[])
    sku = db.StringField(required=False, unique=True)
    slug = db.StringField(required=False, unique=True)
    system_key = db.StringField(required=False, default="")
    description = db.StringField(required=False, default='')
    media = db.ListField(default=[])
    tenant_id = db.StringField(required=False, default="")
    tenant_name = db.StringField(required=False, default="")
    dimension = db.StringField(required=False, default='2d')
    dynamic = db.BooleanField(default=False)
    dynamic_keys = db.ListField(required=False, default=[])
    start_on = db.IntField(required=False, default=0)
    end_on = db.IntField(required=False, default=0)
    generate = db.BooleanField(required=False, default=False)
    dynamic_type = db.StringField(required=False, default="")
    width = db.IntField(min_value=None, max_value=None, default=0)
    maximum_width = db.IntField(min_value=None, max_value=None, default=0)
    minimum_width = db.IntField(min_value=None, max_value=None, default=0)

    height = db.IntField(min_value=None, max_value=None, default=0)
    maximum_height = db.IntField(min_value=None, max_value=None, default=0)
    minimum_height = db.IntField(min_value=None, max_value=None, default=0)

    length = db.IntField(min_value=None, max_value=None, default=0)
    maximum_length = db.IntField(min_value=None, max_value=None, default=0)
    minimum_length = db.IntField(min_value=None, max_value=None, default=0)

    unit = db.StringField(required=False, default='')
    incremental_by = db.IntField(min_value=None, max_value=None, default=0)
    published = db.BooleanField(default=True)
    has_children = db.BooleanField(default=False)
    input_type = db.StringField(default='')
    extended_fields = db.ListField(required=False, default=[])
    shareable = db.BooleanField(default=False)
    parent = db.BooleanField(default=True)
    start_cost = db.IntField(min_value=None, max_value=None, default=0)
    calculation_method = db.ListField(required=False)
    rpm = db.IntField(min_value=None, max_value=None, default=0)
    sheet_runs = db.ListField(required=False, default=[])
    runs = db.ListField(required=False, default=[])
    configure = db.ListField(required=False, default=[])
    information = db.StringField(default='')
    children = db.ListField(db.ReferenceField('self', default=None), default=None)
    additional = db.DictField(required=False, default={})
    created_at = db.DateTimeField(default=datetime.datetime.now, required=True)
    checked = db.BooleanField(default=False)

    meta = {
        'indexes': ['-created_at', 'slug', 'name'],
        'ordering': ['-created_at', 'name'],
        'collection': 'options',
        'queryset_class': BaseQuerySet
    }

    def __setattr__(self, key, value):
        super(Option, self).__setattr__(key, value)
        if key == 'name':
            self.slug = slugify(self.name, to_lower=True)

File: microservices/calculation/calculation/Models/Option.js
const mongoose = require("mongoose");
const Schema = mongoose.Schema;


const OptionSchema = Schema({
    iso: { type: String, required: true },
    sort: { type: Number, required: false, default: 0 },
    name: { type: String, required: true, unique_with: 'tenant_id' },
    display_name: { type: String, required: true },
    slug: { type: String, required: false, unique_with: 'tenant_id' },
    tenant_id: { type: String, required: true },
    tenant_name: { type: String, required: false },
    unit: { type: String, required: false },
    maximum: { type: Number, required: false },
    minimum: { type: Number, required: false },
    incremental_by: { type: Number, required: false },
    information: { type: String, required: false },
    input_type: { type: String, required: false },
    dynamic_keys : {type: Array, required: false, default: []},
    start_on : {type: Number, required: false, default: ''},
    end_on : {type: Number, required: false, default: ''},
    generate : {type: Boolean, required: false, default: ''},
    dynamic_type : {type: String, required: false, default: ''},
    // linked = db.ReferenceField(Option)
    linked: {type:String, required: true, index: true}, // REFER
    shareable: { type: Boolean, required: false, default: false },
    published: { type: Boolean, default: true },
    description: { type: String, required: false, default: '' },
    media: { type: String, default: '' },
    sku: { type: String, default: '' },
    rpm: { type: Number, default: '' },
    sheet_runs: { type: Array, default: '' },
    runs :  { type: Array, default: '' },
    created_at: { type: Date, required: true },
});


module.exports = Option = mongoose.model("options", OptionSchema);
